File: learning/preprocessing/lerp_resizer.py
"""Functions to linearly resample timeseries data.
"""

import numpy as np

EPSILON = 0.000001


def resize(imu_data, sampling_rate, start_time, end_time):
    if start_time < imu_data[0, 0]:
        imu_data = np.insert(imu_data, 0, np.concatenate(([start_time], imu_data[0, 1:])), axis=0)

    if imu_data[-1, 0] < end_time:
        imu_data = np.insert(imu_data, -1, np.concatenate(([end_time], imu_data[-1, 1:])), axis=0)

    dt = end_time - start_time
    l_data = len(imu_data)
    m = int(dt * sampling_rate)
    time_step = 1.0 / sampling_rate
    resized_data = np.zeros(shape=(m, 4))

    t = start_time
    current_index = 0
    for i in range(m):
        # Search the two timestamps around t and interpolate if necessary
        for j in range(current_index, l_data):
            entry = imu_data[j]
            timestamp = entry[0]
            if abs(timestamp - t) < EPSILON:
                # This sample has nearly the same timestamp
                resized_data[i, 0] = t
                resized_data[i, 1:] = entry[1:]
                current_index = j + 1
                break
            elif timestamp > t:
                # Search for entry with timestamp smaller than t
                previous_index = j - 1
                while imu_data[previous_index, 0] > t:
                    previous_index -= 1
                previous_entry = imu_data[previous_index]
                resized_data[i, 0] = t
                resized_data[i, 1:] = interpolate(t, previous_entry, entry)
                current_index = j
                break
        t += time_step
    return resized_data


def interpolate(t, previous_entry, entry):
    t1 = previous_entry[0]
    t2 = entry[0]
    dt = t2 - t1
    w1 = (t2 - t) / dt
    w2 = (t - t1) / dt
    return w1 * previous_entry[1:] + w2 * entry[1:]

File: android/GestEarWear/app/src/main/java/ch/ethz/inf/vs/gestear/datacollection/imu/LerpResizer.java
package ch.ethz.inf.vs.gestear.datacollection.imu;

import java.util.List;

public class LerpResizer {

    private static final float EPSILON = 0.000001f;
    private static LerpResizer instance = null;

    public static LerpResizer getInstance() {
        if (instance == null) {
            instance = new LerpResizer();
        }
        return instance;
    }

    private LerpResizer() {

    }

    public float[][][] resizeSequence(List<IMUEntry> data, int fs, double startTime, float dt, int sequenceLength) {
        double endTime = startTime + dt;

        boolean addedAtStart = false;
        if (startTime < data.get(0).getTimestamp()) {
            data.add(0, new IMUEntry(startTime, data.get(0).getValues()));
            addedAtStart = true;
        }
        boolean addedAtEnd = false;
        if (data.get(data.size() - 1).getTimestamp() < endTime) {
            data.add(new IMUEntry(endTime, data.get(data.size() - 1).getValues()));
            addedAtEnd = true;
        }

        int m = (int) (dt * fs);
        //if (m % 2 == 1) {
        //    m++;
        //}
        float timeStep = 1.0f / fs;
        int windowLength = m / sequenceLength;
        float[][][] result = new float[sequenceLength][windowLength][3];

        double t = startTime;
        int currentIndex = 0;
        for (int i = 0; i < m; i++) {
            // Search the two timestamps around t and interpolate if necessary
            for (int j = currentIndex; j < data.size(); j++) {
                IMUEntry entry = data.get(j);
                double timestamp = entry.getTimestamp();
                if (Math.abs(timestamp - t) < EPSILON) {
                    // This sample has nearly the same timestamp
                    result[i / windowLength][i % windowLength] = entry.getValues();
                    currentIndex = j + 1;
                    break;
                } else if (timestamp > t) {
                    // Search for entry with timestamp smaller than t
                    int previousIndex = j - 1;
                    while (data.get(previousIndex).getTimestamp() > t) {
                        previousIndex--;
                    }
                    IMUEntry previousEntry = data.get(previousIndex);
                    result[i / windowLength][i % windowLength] = interpolate(t, previousEntry.getTimestamp(), previousEntry.getValues(), entry.getTimestamp(), entry.getValues());
                    currentIndex = j;
                    break;
                }
            }
            t += timeStep;
        }

        if (addedAtStart) {
            data.remove(0);
        }
        if (addedAtEnd) {
            data.remove(data.size() - 1);
        }

        return result;
    }

    public float[][] resizeWindow(List<IMUEntry> data, int fs, double startTime, float dt) {
        double endTime = startTime + dt;

        boolean addedAtStart = false;
        if (startTime < data.get(0).getTimestamp()) {
            data.add(0, new IMUEntry(startTime, data.get(0).getValues()));
            addedAtStart = true;
        }
        boolean addedAtEnd = false;
        if (data.get(data.size() - 1).getTimestamp() < endTime) {
            data.add(new IMUEntry(endTime, data.get(data.size() - 1).getValues()));
            addedAtEnd = true;
        }

        int m = (int) (dt * fs);
        //if (m % 2 == 1) {
        //    m++;
        //}
        float timeStep = 1.0f / fs;
        float[][] result = new float[m][3];

        double t = startTime;
        int currentIndex = 0;
        for (int i = 0; i < m; i++) {
            // Search the two timestamps around t and interpolate if necessary
            for (int j = currentIndex; j < data.size(); j++) {
                IMUEntry entry = data.get(j);
                double timestamp = entry.getTimestamp();
                if (Math.abs(timestamp - t) < EPSILON) {
                    // This sample has nearly the same timestamp
                    result[i] = entry.getValues();
                    currentIndex = j + 1;
                    break;
                } else if (timestamp > t) {
                    // Search for entry with timestamp smaller than t
                    int previousIndex = j - 1;
                    while (data.get(previousIndex).getTimestamp() > t) {
                        previousIndex--;
                    }
                    IMUEntry previousEntry = data.get(previousIndex);
                    result[i] = interpolate(t, previousEntry.getTimestamp(), previousEntry.getValues(), entry.getTimestamp(), entry.getValues());
                    currentIndex = j;
                    break;
                }
            }
            t += timeStep;
        }

        if (addedAtStart) {
            data.remove(0);
        }
        if (addedAtEnd) {
            data.remove(data.size() - 1);
        }

        return result;
    }

    private float[] interpolate(double t, double t1, float[] x1, double t2, float[] x2) {
        double dt = t2 - t1;
        float w1 = (float) ((t2 - t) / dt);
        float w2 = (float) ((t - t1) / dt);
        float[] x = new float[x1.length];
        for (int i = 0; i < x.length; i++) {
            x[i] = w1 * x1[i] + w2 * x2[i];
        }
        return x;
    }
}
